python/legate_dataframe/experimental_ibis/__init__.py

```diff
@@ -0,0 +1,309 @@
+            source, delimiter=delimiter, usecols=usecols, dtypes=dtypes
+        )
+
+        # We already have the schema, just insert:
+        self._tables[table_name] = (table, schema)
+        return self.table(table_name)
+
+    def read_parquet(
+        self,
+        source: str | pathlib.Path,
+        table_name: str | None = None,
+    ):
+        if table_name is None:
+            table_name = _gen_name("read_parquet", str(source))
+
+        table = legate_dataframe.lib.parquet.parquet_read(source)
+        return self.create_table(table_name, table)
+
+    def list_tables(self, like=None, database=None):
+        return self._filter_with_like(list(self._tables.keys()), like)
+
+    def table(self, name: str, schema: sch.Schema | None = None):
+        inferred_schema = self.get_schema(name)
+        overridden_schema = {**inferred_schema, **(schema or {})}
+        return ops.DatabaseTable(name, overridden_schema, self).to_expr()
+
+    def get_schema(self, table_name, *, database=None):
+        _, schema = self._tables[table_name]
+
+        return schema
+
+    def compile(self, expr, *args, **kwargs):
+        # We don't have a different intermediate representation, so do nothing
+        return expr
+
+    def create_table(
+        self,
+        name: str,
+        obj=None,
+        *,
+        schema: sch.Schema | None = None,
+        database: str | None = None,
+        temp: bool | None = None,
+        overwrite: bool = False,
+    ) -> ir.Table:
+        """Create a table."""
+        if temp is not None and not temp:
+            com.IbisError(
+                "Passing `temp=False` to the Legate backend's `create_table()` method "
+                "is not supported: all tables are in memory and temporary."
+            )
+
+        if database is not None:
+            com.IbisError(
+                "Passing `database` to the Legate backend's `create_table()` method "
+                "is not supported: Legate cannot set a database."
+            )
+
+        if name in self._tables and not overwrite:
+            raise com.IbisError(f"Cannot overwrite existing table `{name}`")
+
+        if obj is None and schema is None:
+            raise com.IbisError("The schema or obj parameter is required")
+
+        if schema is not None:
+            schema = ibis.schema(schema)
+
+        if obj is not None:
+            table = self._convert_object(obj)
+            if schema is None:
+                schema = infer_schema_from_logical_table(table)
+            else:
+                # Check if schema matches actual table (is it right to not
+                # enforce the schema by dropping cols or casting?)
+                actual_schema = infer_schema_from_logical_table(table)
+                if schema != actual_schema:
+                    # TODO: General Q: when to use Ibis errors and when not?!
+                    # (I find the broad use of IbisError a bit surprising.)
```

Comment:
This (on the Ibis side) is ongoing work.

In the (very soon to be released) 10.0, we unified all of the various ways that different backends report failures under a common `TableNotFound` exception.  

But we do tend towards `IbisError` for user interactions, especially where passing through the backend error would lead to different exception types depending on the backend being used.

Reply:
Sounds like this is fine, and keeping `NotImplementedError` also where it is used  (will just remove the comment for now).

(Specific errors, as long as there are not too many, does seem nice.)